### PR TITLE
RBAC Authorization Engine updates based on A41 proposal

### PR DESCRIPTION
**Pull request by ashithasantosh**
Proposal: https://github.com/grpc/proposal/blob/master/A41-xds-rbac.md
1. `remote_ip` is the same as `direct_remote_ip` as long as ProxyProtocol and xff_num_trusted_hops are unsupported.
2. `requested_server_name` can match if the matcher accepts empty string.
3. Update `HeaderMatcher` to return failed match when value is not present and match_pattern is not present match.
4. Support for `MetadataMatcher` field in RBAC config.
5. Maps `host` to `:authority` when fetching value from metadata.
6. Treats `TE` header as not present.

---
This change is [Reviewable](https://reviewable.io/reviews/grpc/grpc/27754)

### CODE REVIEW

**Comment by ashithasantosh** [comment]
From A41:xDS RBAC Support
`In RBAC metadata refers to the Envoy metadata which has no relation to gRPC metadata. Envoy metadata is generic state shared between filters, which has no gRPC equivalent. RBAC implementations in gRPC will treat Envoy metadata as an empty map. Since ValueMatcher can only match if a value is present (even NullMatch), the metadata matcher is guaranteed not to match.`

As I understand, if we have a rule with MetadataMatcher mentioned, that will always return false since we are matching against an empty map, so any [match_pattern](https://github.com/envoyproxy/envoy/blob/53f95570a7e9d7fa5414a4bc145852524fa017ed/api/envoy/type/matcher/v3/value.proto#L32) would give failed match, irrespective of the contents in MetadataMatcher field in RBAC config. Except for [invert](https://github.com/envoyproxy/envoy/blob/main/api/envoy/type/matcher/v3/metadata.proto#L106) field, which if true, would result in successful match.

@markdroth @ejona86
Please confirm if my understanding is correct. Then I will go ahead and make the required changes. Thank you!

**Comment by ejona86** [comment]
> The only value we check for is the invert field, which if true, would result in successful match.

Invert doesn't change the presence check. I think it uses ternary logic. I believe not_rule-ing the matcher is the only method of producing a match. It is possible that may vary for metadata vs (say) header matchers.

**Comment by ashithasantosh** [comment]
> Invert doesn't change the presence check.

I think invert does not depend on the match pattern type, it simply inverts the match result. Similar to not_rule.
https://github.com/envoyproxy/envoy/blob/9d3cf43e3fab618dd2bedd0deb2df973760bbbd6/source/common/common/matchers.cc#L121

And present_match would always return false, since the value never exists.
https://github.com/envoyproxy/envoy/blob/9d3cf43e3fab618dd2bedd0deb2df973760bbbd6/source/common/common/matchers.cc#L48

**Comment by ejona86** [comment]
For headers, it mostly uses ternary logic (except for presence match):
https://github.com/envoyproxy/envoy/blob/b7393b0bd221dab43bea0d697df0954a5b01d7d7/source/common/http/header_utility.cc#L140

**Comment by markdroth** [comment]
> > For headers, it mostly uses ternary logic (except for presence match): https://github.com/envoyproxy/envoy/blob/b7393b0bd221dab43bea0d697df0954a5b01d7d7/source/common/http/header_utility.cc#L140
> 
> Right! C++ code https://github.com/grpc/grpc/blob/master/src/core/lib/matchers/matchers.cc#L287 I do have to update the implementation to return failed match for match types other than `present` when value is not present. Will add that to this PR. Thanks for bringing this to my attention.:)

I think the change you made to the header present match semantics is incorrect.  Our implementation was already doing the right thing, and Envoy's behavior was recently changed to match ours.  For discussion, see envoyproxy/envoy#14656.  Envoy's behavior was fixed to match gRPC's in envoyproxy/envoy#16627.

@ejona86, please make sure this matches what you expected, and is correct in all languages.

File: src/core/lib/matchers/matchers.cc [hunk 1]
```diff
@@ -291,7 +291,7 @@ bool HeaderMatcher::Match(
     match = value.has_value() == present_match_;
   } else if (!value.has_value()) {
     // All other types fail to match if field is not present.
-    match = false;
+    return false;
```
Comment (ashithasantosh) [inline src/core/lib/matchers/matchers.cc:294]: This block would be triggered for types that are not `Present match` and value is absent.
Envoy also returns failed match in that case, irrespective of the invert value.
https://github.com/envoyproxy/envoy/blob/4c76813087c07171996d2cd8a9c591604b9124d8/source/common/http/header_utility.cc#L142
@markdroth @ejona86 Please correct me if I am wrong. Thank you.

Comment (markdroth) [reply src/core/lib/matchers/matchers.cc:294]: Sorry, you're right -- I misread the code change you made.  I somehow thought you were changing the behavior of present_match to always return false, regardless of the value of invert_match.

Sorry for the confusion!

**Review by markdroth — CHANGES_REQUESTED** [review]
Sorry for taking so long to review this.  It looks good overall!  Comments are mostly minor, but there are a couple of things that we should make sure to get consistent between this and the work that @yashykt is doing.

Please let me know if you have any questions.  Thanks!

> Reviewed 9 of 9 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/27754#-:-Mq_zF3G9Vclv70Vdkrr:by24x7e)* status: all files reviewed, 5 unresolved discussions (waiting on @ashithasantosh)

___
*[src/core/lib/security/authorization/evaluate_args.cc, line 120 at r2](https://reviewable.io/reviews/grpc/grpc/27754#-Mq_JNmK2w9n_Ag9ROJc:-Mq_JNmL2LYC8mJz_vqg:biuhsb3) ([raw file](https://github.com/grpc/grpc/blob/d8265dbe0897cc79f3b5b82953b93011f5a2698e/src/core/lib/security/authorization/evaluate_args.cc#L120)):*
> ```cpp
>     return absl::nullopt;
>   }
>   if (absl::EqualsIgnoreCase(key, "te")) {
> ```

Do we want to have these special cases here, or should we handle them in the RBAC filter itself?  @yashykt 
___
*[src/core/lib/security/authorization/evaluate_args.cc, line 124 at r2](https://reviewable.io/reviews/grpc/grpc/27754#-Mq_JTK759u2kSn9VwG5:-Mq_JTK834VgFFUkBGsb:b-jmj8xo) ([raw file](https://github.com/grpc/grpc/blob/d8265dbe0897cc79f3b5b82953b93011f5a2698e/src/core/lib/security/authorization/evaluate_args.cc#L124)):*
> ```cpp
>   }
>   if (absl::EqualsIgnoreCase(key, "host")) {
>     // Maps legacy host header to :authority.
> ```

@yashykt Does this match the behavior in the chttp2 code?  I believe we decided that the transport behavior should be:
- Fail the call if there are duplicate `Host` or `:authority` headers.
- If both `Host` and `:authority` are set, drop `Host`.
- If `Host` is set but not `:authority`, rename it to `:authority`.

@ejona86 to keep me honest here.
___
*[src/core/lib/security/authorization/evaluate_args.cc, line 127 at r2](https://reviewable.io/reviews/grpc/grpc/27754#-Mq_E-J03JHr5jA-zEeh:-Mq_E-J19M6221hCGkMk:b-xml9ne) ([raw file](https://github.com/grpc/grpc/blob/d8265dbe0897cc79f3b5b82953b93011f5a2698e/src/core/lib/security/authorization/evaluate_args.cc#L127)):*
> ```cpp
>     return metadata_->GetValue(":authority", concatenated_value);
>   }
>   if (absl::EqualsIgnoreCase(key, ":method")) {
> ```

I don't think hard-coding this to "POST" is what we want.  @yashykt has code in #28309 that is setting the `:method` header appropriate in the metadata batch that is passed to `EvaluateArgs`, so I suspect that we can just remove this.
___
*[src/core/lib/security/authorization/rbac_policy.h, line 80 at r2](https://reviewable.io/reviews/grpc/grpc/27754#-Mq_L-6G6Cap3EzbptBv:-Mq_L-6G6Cap3EzbptBw:b-5wde8g) ([raw file](https://github.com/grpc/grpc/blob/d8265dbe0897cc79f3b5b82953b93011f5a2698e/src/core/lib/security/authorization/rbac_policy.h#L80)):*
> ```objc
>     // For kMetadata RuleType. All the other fields in MetadataMatcher are
>     // ignored except invert.
>     Permission(Permission::RuleType type, bool invert);
> ```

I'm not sure this ctor is actually distinct from the previous one, since `bool` is convertible to `int`.  I worry that the compiler may choose the wrong one in some cases.

A better pattern here might be to have a different factory method for each rule type.  For example:

```
static Permission MakeDestPortPermission(int port);
static Permission MakeMetadataPermission(bool invert);
// ...etc...
```

That way, the caller can explicitly indicate which type it wants.
___
*[src/core/lib/security/authorization/rbac_policy.h, line 130 at r2](https://reviewable.io/reviews/grpc/grpc/27754#-Mq_LtV_6cbfODABMxLS:-Mq_LtV_6cbfODABMxLT:biynud4) ([raw file](https://github.com/grpc/grpc/blob/d8265dbe0897cc79f3b5b82953b93011f5a2698e/src/core/lib/security/authorization/rbac_policy.h#L130)):*
> ```objc
>     // For kMetadata RuleType. All the other fields in MetadataMatcher are
>     // ignored except invert.
>     Principal(Principal::RuleType type, bool invert);
> ```

The same factory-method approach I suggest above would probably be a good idea here too, even though there's no issue of the compiler choosing the wrong ctor in this particular case.

**Review by ashithasantosh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/27754)* status: 8 of 9 files reviewed, 5 unresolved discussions (waiting on @ashithasantosh, @markdroth, and @yashykt)

___
*[src/core/lib/security/authorization/evaluate_args.cc, line 120 at r2](https://reviewable.io/reviews/grpc/grpc/27754#-Mq_JNmK2w9n_Ag9ROJc:-MquUa48EZuXn1ijjJvx:b-5uq6m0) ([raw file](https://github.com/grpc/grpc/blob/d8265dbe0897cc79f3b5b82953b93011f5a2698e/src/core/lib/security/authorization/evaluate_args.cc#L120)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Do we want to have these special cases here, or should we handle them in the RBAC filter itself?  @yashykt 
</blockquote></details>

Anything works.
From the proposal, it seemed like something we may want to do in the common code (RBAC matchers).
`Since hop-by-hop headers are not used in HTTP/2 except for te: trailers, transports must consider requests containing the Connection header as malformed, independent of xDS or RBAC, and only the TE header may need special handling. If the transport exposes TE in Metadata, then RBAC must special-case the header to treat it as not present. `
___
*[src/core/lib/security/authorization/evaluate_args.cc, line 127 at r2](https://reviewable.io/reviews/grpc/grpc/27754#-Mq_E-J03JHr5jA-zEeh:-MquWtUQ6SaLtC9uNv4B:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/d8265dbe0897cc79f3b5b82953b93011f5a2698e/src/core/lib/security/authorization/evaluate_args.cc#L127)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

I don't think hard-coding this to "POST" is what we want.  @yashykt has code in #28309 that is setting the `:method` header appropriate in the metadata batch that is passed to `EvaluateArgs`, so I suspect that we can just remove this.
</blockquote></details>

Done.
___
*[src/core/lib/security/authorization/rbac_policy.h, line 80 at r2](https://reviewable.io/reviews/grpc/grpc/27754#-Mq_L-6G6Cap3EzbptBv:-MquWxVI3PEVZVLMTZRy:bny1kkt) ([raw file](https://github.com/grpc/grpc/blob/d8265dbe0897cc79f3b5b82953b93011f5a2698e/src/core/lib/security/authorization/rbac_policy.h#L80)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

I'm not sure this ctor is actually distinct from the previous one, since `bool` is convertible to `int`.  I worry that the compiler may choose the wrong one in some cases.

A better pattern here might be to have a different factory method for each rule type.  For example:

```
static Permission MakeDestPortPermission(int port);
static Permission MakeMetadataPermission(bool invert);
// ...etc...
```

That way, the caller can explicitly indicate which type it wants.
</blockquote></details>

Unless the caller uses an int instead of bool or other way around, I don't think the compiler would choose the wrong function.
Plus when we use the factory method, we would be calling ctor internally, and that could cause the same problem.
I was thinking of maybe adding assert statements (ex. for type) in ctor body, to confirm the right ctor is being invoked. Wdyt?

**Review by markdroth — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/27754)* status: 8 of 9 files reviewed, 5 unresolved discussions (waiting on @ashithasantosh, @markdroth, and @yashykt)

___
*[src/core/lib/security/authorization/rbac_policy.h, line 80 at r2](https://reviewable.io/reviews/grpc/grpc/27754#-Mq_L-6G6Cap3EzbptBv:-MquhdjG2sXihXSEUXzw:b-lccwbm) ([raw file](https://github.com/grpc/grpc/blob/d8265dbe0897cc79f3b5b82953b93011f5a2698e/src/core/lib/security/authorization/rbac_policy.h#L80)):*
<details><summary><i>Previously, ashithasantosh (Ashitha Santhosh) wrote…</i></summary><blockquote>

Unless the caller uses an int instead of bool or other way around, I don't think the compiler would choose the wrong function.
Plus when we use the factory method, we would be calling ctor internally, and that could cause the same problem.
I was thinking of maybe adding assert statements (ex. for type) in ctor body, to confirm the right ctor is being invoked. Wdyt? 
</blockquote></details>

The point is that `bool` is convertable to `int`, and we still have a lot of C-style code that uses `int` instead of `bool`.  I think that even if the compiler always gets the right one for the type being passed in, it's not always going to be true that the right type will be passed in.  And this API has no protection against the wrong thing happening.

I would really prefer to use factory method pattern here, since it would be semantically very precise and difficult to misuse.

**Review by ashithasantosh — COMMENTED** [review]
Thank you for the review:) I made the suggested changes.

> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/27754#-:-MqvMsZ_9LyrKJ0_LJx6:b-aom9zg)* status: 5 of 11 files reviewed, 5 unresolved discussions (waiting on @ashithasantosh, @markdroth, and @yashykt)

___
*[src/core/lib/security/authorization/rbac_policy.h, line 80 at r2](https://reviewable.io/reviews/grpc/grpc/27754#-Mq_L-6G6Cap3EzbptBv:-MqvLm_0EAINqbAHqmY6:b-nmvcas) ([raw file](https://github.com/grpc/grpc/blob/d8265dbe0897cc79f3b5b82953b93011f5a2698e/src/core/lib/security/authorization/rbac_policy.h#L80)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

The point is that `bool` is convertable to `int`, and we still have a lot of C-style code that uses `int` instead of `bool`.  I think that even if the compiler always gets the right one for the type being passed in, it's not always going to be true that the right type will be passed in.  And this API has no protection against the wrong thing happening.

I would really prefer to use factory method pattern here, since it would be semantically very precise and difficult to misuse.
</blockquote></details>

Sure thing. Updated code to use factory methods.
___
*[src/core/lib/security/authorization/rbac_policy.h, line 130 at r2](https://reviewable.io/reviews/grpc/grpc/27754#-Mq_LtV_6cbfODABMxLS:-MqvM0Aq7oq0jfv71EvW:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/d8265dbe0897cc79f3b5b82953b93011f5a2698e/src/core/lib/security/authorization/rbac_policy.h#L130)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

The same factory-method approach I suggest above would probably be a good idea here too, even though there's no issue of the compiler choosing the wrong ctor in this particular case.
</blockquote></details>

Done.

**Review by markdroth — CHANGES_REQUESTED** [review]
This looks good!  The only thing left is to get feedback from @yashykt on the questions I tagged him on.

Please let me know if you have any questions.  Thanks!

> Reviewed 1 of 1 files at r3, 3 of 5 files at r4, 6 of 6 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/27754#-:-MqzSa7i9kVnjJW_iVbs:b-6tg1ia)* status: all files reviewed, 5 unresolved discussions (waiting on @ashithasantosh and @yashykt)

___
*[src/core/lib/security/authorization/rbac_policy.h, line 49 at r5](https://reviewable.io/reviews/grpc/grpc/27754#-MqzRQ-z4cEyfzHSff2U:-MqzRQ0-0Uz9SREiUyzh:b-cuo876) ([raw file](https://github.com/grpc/grpc/blob/90c1965c81575e5bd5b3b4cb08d8006f70cea44d/src/core/lib/security/authorization/rbac_policy.h#L49)):*
> ```objc
>   // TODO(ashithasantosh): Support for destination_port_range.
>   struct Permission {
>    public:
> ```

No need for access control, since this is a struct instead of a class.  It's fine for everything to be public.
___
*[src/core/lib/security/authorization/rbac_policy.h, line 97 at r5](https://reviewable.io/reviews/grpc/grpc/27754#-MqzRYAC5c6MLv29HnNC:-MqzRYAC5c6MLv29HnND:b-9e6vne) ([raw file](https://github.com/grpc/grpc/blob/90c1965c81575e5bd5b3b4cb08d8006f70cea44d/src/core/lib/security/authorization/rbac_policy.h#L97)):*
> ```objc
>    private:
>     // For kAnd/kOr RuleType.
>     Permission(Permission::RuleType type,
> ```

These ctors are probably not needed at all anymore.  The factory methods can just fill in the fields directly.
___
*[src/core/lib/security/authorization/rbac_policy.h, line 116 at r5](https://reviewable.io/reviews/grpc/grpc/27754#-MqzRlfV7N0C4jf2u_xY:-MqzRlfV7N0C4jf2u_xZ:b7mx7ln) ([raw file](https://github.com/grpc/grpc/blob/90c1965c81575e5bd5b3b4cb08d8006f70cea44d/src/core/lib/security/authorization/rbac_policy.h#L116)):*
> ```objc
> 
>   struct Principal {
>    public:
> ```

Same as above: no need for access control, and no need for ctors.

**Review by ashithasantosh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/27754)* status: all files reviewed, 5 unresolved discussions (waiting on @ashithasantosh, @markdroth, and @yashykt)

___
*[src/core/lib/security/authorization/rbac_policy.h, line 49 at r5](https://reviewable.io/reviews/grpc/grpc/27754#-MqzRQ-z4cEyfzHSff2U:-Mr1Ry-Y95Tb933uJLYS:b-cwvma5) ([raw file](https://github.com/grpc/grpc/blob/90c1965c81575e5bd5b3b4cb08d8006f70cea44d/src/core/lib/security/authorization/rbac_policy.h#L49)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

No need for access control, since this is a struct instead of a class.  It's fine for everything to be public.
</blockquote></details>

I added the access control to avoid exposing constructors. So that users only use the factory methods.
___
*[src/core/lib/security/authorization/rbac_policy.h, line 97 at r5](https://reviewable.io/reviews/grpc/grpc/27754#-MqzRYAC5c6MLv29HnNC:-Mr1Sd9x-qz1oCRec6tx:bgm6q84) ([raw file](https://github.com/grpc/grpc/blob/90c1965c81575e5bd5b3b4cb08d8006f70cea44d/src/core/lib/security/authorization/rbac_policy.h#L97)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

These ctors are probably not needed at all anymore.  The factory methods can just fill in the fields directly.
</blockquote></details>

Since the methods are static, the data members would need to be static as well. That's why we need the constructors.

**Review by markdroth — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/27754)* status: all files reviewed, 5 unresolved discussions (waiting on @ashithasantosh, @markdroth, and @yashykt)

___
*[src/core/lib/security/authorization/rbac_policy.h, line 97 at r5](https://reviewable.io/reviews/grpc/grpc/27754#-MqzRYAC5c6MLv29HnNC:-Mr2klGM8LRQc9dpNURG:b6hn4c1) ([raw file](https://github.com/grpc/grpc/blob/90c1965c81575e5bd5b3b4cb08d8006f70cea44d/src/core/lib/security/authorization/rbac_policy.h#L97)):*
<details><summary><i>Previously, ashithasantosh (Ashitha Santhosh) wrote…</i></summary><blockquote>

Since the methods are static, the data members would need to be static as well. That's why we need the constructors.
</blockquote></details>

You don't need the constructors.  The static methods can directly access the data members of the new object:

```
Rbac::Permission Rbac::Permission::MakeDestIpPermission(CidrRange ip) {
  Permission permission;
  permission.type = RuleType::kDestIp;
  permission.ip = std::move(ip);
  return permission;
}
```

**Review by ashithasantosh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/27754)* status: 10 of 12 files reviewed, 5 unresolved discussions (waiting on @ashithasantosh, @markdroth, and @yashykt)

___
*[src/core/lib/security/authorization/rbac_policy.h, line 97 at r5](https://reviewable.io/reviews/grpc/grpc/27754#-MqzRYAC5c6MLv29HnNC:-Mr3eJbO8e_YrT6Y8UMi:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/90c1965c81575e5bd5b3b4cb08d8006f70cea44d/src/core/lib/security/authorization/rbac_policy.h#L97)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

You don't need the constructors.  The static methods can directly access the data members of the new object:

```
Rbac::Permission Rbac::Permission::MakeDestIpPermission(CidrRange ip) {
  Permission permission;
  permission.type = RuleType::kDestIp;
  permission.ip = std::move(ip);
  return permission;
}
```
</blockquote></details>

Done.
___
*[src/core/lib/security/authorization/rbac_policy.h, line 116 at r5](https://reviewable.io/reviews/grpc/grpc/27754#-MqzRlfV7N0C4jf2u_xY:-Mr3eRwQ2Fuxb3RQ2xIA:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/90c1965c81575e5bd5b3b4cb08d8006f70cea44d/src/core/lib/security/authorization/rbac_policy.h#L116)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Same as above: no need for access control, and no need for ctors.
</blockquote></details>

Done.

**Review by markdroth — COMMENTED** [review]
The code looks good!  Just need confirmation from @yashykt on the questions I tagged him on.

> Reviewed 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/27754#-:-MsakKEaAbVmd1REIkdI:b-bh3xnk)* status: all files reviewed, 2 unresolved discussions (waiting on @ashithasantosh and @yashykt)

**Review by yashykt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/27754)* status: 9 of 12 files reviewed, 2 unresolved discussions (waiting on @ejona86 and @markdroth)

___
*[src/core/lib/security/authorization/evaluate_args.cc, line 120 at r2](https://reviewable.io/reviews/grpc/grpc/27754#-Mq_JNmK2w9n_Ag9ROJc:-Msff5_-BP6lZi2sOAwF:b-85i649) ([raw file](https://github.com/grpc/grpc/blob/d8265dbe0897cc79f3b5b82953b93011f5a2698e/src/core/lib/security/authorization/evaluate_args.cc#L120)):*
<details><summary><i>Previously, ashithasantosh (Ashitha Santhosh) wrote…</i></summary><blockquote>

Anything works.
From the proposal, it seemed like something we may want to do in the common code (RBAC matchers).
`Since hop-by-hop headers are not used in HTTP/2 except for te: trailers, transports must consider requests containing the Connection header as malformed, independent of xDS or RBAC, and only the TE header may need special handling. If the transport exposes TE in Metadata, then RBAC must special-case the header to treat it as not present. `
</blockquote></details>

LGTM 
___
*[src/core/lib/security/authorization/evaluate_args.cc, line 124 at r2](https://reviewable.io/reviews/grpc/grpc/27754#-Mq_JTK759u2kSn9VwG5:-MsfgFsu2croojlbv1t4:b-3sxkrm) ([raw file](https://github.com/grpc/grpc/blob/d8265dbe0897cc79f3b5b82953b93011f5a2698e/src/core/lib/security/authorization/evaluate_args.cc#L124)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

@yashykt Does this match the behavior in the chttp2 code?  I believe we decided that the transport behavior should be:
- Fail the call if there are duplicate `Host` or `:authority` headers.
- If both `Host` and `:authority` are set, drop `Host`.
- If `Host` is set but not `:authority`, rename it to `:authority`.

@ejona86 to keep me honest here.
</blockquote></details>

LGTM
(http code- https://github.com/grpc/grpc/blob/e520fc137c5550d29921d8bf079cfd4616bc4088/src/core/ext/filters/http/server/http_server_filter.cc#L223)

**Review by ashithasantosh — COMMENTED** [review]
Thank you @yashykt for answering the questions. 
@markdroth all the comments are resolved. PTAL. Thank you!

> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/27754#-:-MsltMWj4BqAY0WBEHSb:b-8m442n)* status: 8 of 14 files reviewed, 2 unresolved discussions (waiting on @ejona86 and @markdroth)

**Review by yashykt — APPROVED** [review]

**Review by markdroth — APPROVED** [review]
Sorry for the delay on this!

There are a couple of questions that we need to follow up on, but neither of them need to block this PR.

Thanks!

> Reviewed 1 of 3 files at r7, 3 of 4 files at r8, 3 of 3 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/27754#-:-Mt4wyl14tuXF_TPFQll:b-uydar8)* status: all files reviewed, 2 unresolved discussions (waiting on @ashithasantosh and @ejona86)

___
*[src/core/lib/security/authorization/evaluate_args.cc, line 124 at r2](https://reviewable.io/reviews/grpc/grpc/27754#-Mq_JTK759u2kSn9VwG5:-Mt4uyYqCOmDm8z2O_Sm:b-isnbpf) ([raw file](https://github.com/grpc/grpc/blob/d8265dbe0897cc79f3b5b82953b93011f5a2698e/src/core/lib/security/authorization/evaluate_args.cc#L124)):*
<details><summary><i>Previously, yashykt (Yash Tibrewal) wrote…</i></summary><blockquote>

LGTM
(http code- https://github.com/grpc/grpc/blob/e520fc137c5550d29921d8bf079cfd4616bc4088/src/core/ext/filters/http/server/http_server_filter.cc#L223)
</blockquote></details>

That code takes care of the last bullet in my list.  Do we also properly handle the first two bullets?

This remaining investigation does not need to block this PR, but we do need to figure this out before we de-experimentalize the RBAC functionality.
___
*[src/core/lib/security/authorization/evaluate_args.cc, line 93 at r9](https://reviewable.io/reviews/grpc/grpc/27754#-Mt4w0DI9QnqqwAmdYYJ:-Mt4w0DI9QnqqwAmdYYK:b-7flrmq) ([raw file](https://github.com/grpc/grpc/blob/425d3d5d1e5ea868cc738063c1b17c399e8c64fc/src/core/lib/security/authorization/evaluate_args.cc#L93)):*
> ```cpp
> }
> 
> absl::string_view EvaluateArgs::GetHost() const {
> ```

Do we actually need this method anymore?  It seems like we should always use `GetAuthority()` instead.

This doesn't need to block this PR, but please make sure to follow up on this.

**Review by ashithasantosh — COMMENTED** [review]
Thank you for the approval:)

> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/27754#-:-Mt4yGfq7Hc7pa0JIalc:bvwygc4)* status: all files reviewed, 2 unresolved discussions (waiting on @ejona86 and @markdroth)

___
*[src/core/lib/security/authorization/evaluate_args.cc, line 93 at r9](https://reviewable.io/reviews/grpc/grpc/27754#-Mt4w0DI9QnqqwAmdYYJ:-Mt4yPhy0MJaVOlLl_s7:be0l4bi) ([raw file](https://github.com/grpc/grpc/blob/425d3d5d1e5ea868cc738063c1b17c399e8c64fc/src/core/lib/security/authorization/evaluate_args.cc#L93)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Do we actually need this method anymore?  It seems like we should always use `GetAuthority()` instead.

This doesn't need to block this PR, but please make sure to follow up on this.
</blockquote></details>

Yeah, we don't need this method anymore. Will create a follow-up PR to clean-up EvaluateArgs.

**Pull request merged**